test-framework/src/test/java/org/apache/brooklyn/test/framework/TestEndpointReachableTest.java

```diff
@@ -0,0 +1,222 @@
+        app.sensors().set(sensor, sensorVal);
+        app.start(ImmutableList.of(loc));
+    }
+
+    // Not a normal unit-test, because other processes on shared infrastructure can grab the port between
+    // us identifying it is free and us checking the reachability.
+    @Test(groups="Integration")
+    public void testHardcodedEndpointNotReachable() throws Exception {
+        HostAndPort unusedPort = findUnusedPort();
+        TestEndpointReachable test = app.createAndManageChild(EntitySpec.create(TestEndpointReachable.class)
+                .configure(TestEndpointReachable.TARGET_ENTITY, app)
+                .configure(TestEndpointReachable.ENDPOINT, unusedPort.toString())
+                .configure(TestEndpointReachable.TIMEOUT, Duration.millis(100)));
+
+        try {
+            app.start(ImmutableList.of(loc));
+        } catch (Exception e) {
+            // It should fail to start
+            Exceptions.propagateIfFatal(e);
+            LOG.debug("As desired, failed to start app with TestEndpointReachable: "+e.toString());
+        }
+        EntityAsserts.assertAttributeEqualsEventually(test, TestEndpointReachable.SERVICE_UP, false);
+        EntityAsserts.assertAttributeEqualsEventually(test, Attributes.SERVICE_STATE_ACTUAL, Lifecycle.ON_FIRE);
+    }
+
+    // Not a normal unit-test, because other processes on shared infrastructure can grab the port between
+    // us identifying it is free and us checking the reachability.
+    @Test(groups="Integration")
+    public void testSensorEndpointNotReachable() throws Exception {
+        AttributeSensor<String> sensor = Sensors.newStringSensor("test.reachable.endpoint");
+        HostAndPort unusedPort = findUnusedPort();
+        TestEndpointReachable test = app.createAndManageChild(EntitySpec.create(TestEndpointReachable.class)
+                .configure(TestEndpointReachable.TARGET_ENTITY, app)
+                .configure(TestEndpointReachable.ENDPOINT, unusedPort.toString())
+                .configure(TestEndpointReachable.TIMEOUT, Duration.millis(100)));
+        app.sensors().set(sensor, unusedPort.toString());
+        
+        try {
+            app.start(ImmutableList.of(loc));
+        } catch (Exception e) {
+            // It should fail to start
+            Exceptions.propagateIfFatal(e);
+            LOG.debug("As desired, failed to start app with TestEndpointReachable: "+e.toString());
+        }
+        EntityAsserts.assertAttributeEqualsEventually(test, TestEndpointReachable.SERVICE_UP, false);
+        EntityAsserts.assertAttributeEqualsEventually(test, Attributes.SERVICE_STATE_ACTUAL, Lifecycle.ON_FIRE);
+    }
+
+    protected ServerSocket openServerPort() {
+        int startPort = 58767;
+        int endPort = 60000;
+        int port = startPort;
+        InetAddress localAddress = Networking.getLocalHost();
+        do {
+            try {
+                ServerSocket ss = new ServerSocket(port, 1024, localAddress);
```

Comment:
Passing port `0` will let the OS allocate an available port - no need for the retry logic.